Sprint 15

MickeyRU opened the pull request.

MickeyRU committed:
Фикс бага с скролом тейбл вью с опциями для трекера. Добавлены эмоджи на экран создания трекера.

MickeyRU committed:
Реализован функционал Самостоятельной задачи:

На экран создания трекера добавлено:
коллекция для выбора эмоджи (выбрать можно только одну эмоджи; выделяем её серым фоном);
коллекция для выбора цвета (цвет тоже можно выбрать только один; выделяем его тонким обрамлением);

MickeyRU committed:
Исправлена работа экрана создания трекера - добавлен единственный выбор цвета и эмоджи и их использование при сохранении привычки

MickeyRU committed:
Добавлен датапровайдер и Store для трех сущностей + Core Data

MickeyRU committed:
Написана основа по ТЗ - Создана БД и классы для работы с БД

MickeyRU committed:
Написал функционал по загрузке данных из CoreData при первом запуске приложения +  при изменениях в fetchController-ре через делегаты. Работоспособность кода не проверял.

MickeyRU committed:
Реализован функционал coredata - можно сохранять трекеры и работает фильтрация. Доработать функционал выполнения трекеров по дням!

MickeyRU committed:
Реализован весь функционал 15 спринта.

LizaIlinykh commented on Tracker/Application/AppDelegate.swift:
```diff
@@ -6,9 +6,20 @@
 
 @main
 class AppDelegate: UIResponder, UIApplicationDelegate {
+    lazy var persistentContainer: NSPersistentContainer = {
```
это свойство нужно заприватить для установки (private(set)). у нас не должно быть возможности установить это свойство извне

LizaIlinykh commented on Tracker/Screens/ColorCell.swift:
```diff
@@ -0,0 +1,52 @@
+final class ColorCell: UICollectionViewCell {
+    static let reuseIdentifier = "ColorCell"
+    
+    private var colorView: UIView = {
```
colorView можно сделать let раз она не изменяется

LizaIlinykh commented on Tracker/Screens/ColorCell.swift:
```diff
@@ -0,0 +1,52 @@
+    }
+    
+    func colorIsSelected(isSelected: Bool) {
+        if isSelected == true {
```
здесь можно просто if isSelected  { 
также предлагаю по другому назвать метод, например didSelectColor
название  isColorSelected подойдет методу, который возвращает булевое значение

LizaIlinykh commented on Tracker/Screens/ColorsHelper.swift:
```diff
@@ -0,0 +1,25 @@
+import UIKit
+
+final class ColorsHelper {
+    static let shared = ColorsHelper()
```
если создаешь синглтон, то лучше приватить инициализатор
https://medium.com/@alessandromanilii/singleton-pattern-in-swift-5c7acde96c3e

LizaIlinykh commented on Tracker/Screens/ColorsHelper.swift:
```diff
@@ -0,0 +1,25 @@
+final class ColorsHelper {
+    static let shared = ColorsHelper()
+    
+    func GenerateColors() -> [UIColor] {
```
методы по код стайлу именуем с маленькой буквы =)

LizaIlinykh commented on Tracker/Screens/DataProvider.swift:
```diff
@@ -0,0 +1,176 @@
+    private let context: NSManagedObjectContext
+    private let trackerStore: TrackerStoreProtocol
+    private let trackerCategoryStore: TrackerCategoryStoreProtocol
+    private let trackerRecordsStore: TrackerRecordStoreProtocol
```
отлично, что все закрыто протоколами и зависимости внедрены через инициализатор! =)

LizaIlinykh commented on Tracker/Screens/EmojiCell.swift:
```diff
@@ -0,0 +1,51 @@
+final class EmojiCell: UICollectionViewCell {
+    static let reuseIdentifier = "EmojiCell"
+    
+    private var emojiTextLabel: UILabel = {
```
emojiTextLabel можно сделать let

LizaIlinykh commented on Tracker/Screens/EmojiCell.swift:
```diff
@@ -0,0 +1,51 @@
+        self.emojiTextLabel.text = emoji
+    }
+    
+    func emojiIsSelected(isSelected: Bool) {
```
здесь тоже советую переименовать метод

LizaIlinykh commented on Tracker/Screens/HeaderViewForEmojiAndColor.swift:
```diff
@@ -0,0 +1,32 @@
+import UIKit
+
+final class HeaderViewForEmojiAndColor: UICollectionReusableView {
+    let titleLabel: UILabel = {
```
titleLabel нужно сделать приватным и сделать метод configureCell для настройки view с определенным текстом, точно также как сделано у ячеек

LizaIlinykh commented on Tracker/Screens/TrackersViewController.swift:
```diff
@@ -8,17 +8,13 @@
-    // Категории для отображения в UI
-    private var visibleCategories: [TrackerCategory] = []
-    // Выполненные трекеры
+    private var dataProvider: DataProviderProtocol!
```
необходимо убрать forceunwrap у dataProvider
dataProvider должен быть передан в инициализаторе TrackersViewController

LizaIlinykh commented on Tracker/Screens/TrackersViewController.swift:
```diff
@@ -8,17 +8,13 @@
     private var completedTrackers: [TrackerRecord] = []
     // Текущая дата
-    private var currentDate: Date
+    private var currentDate: Date!
```
необходимо убрать forceunwrap у currentDate, можно заменить на обычный опционал

LizaIlinykh commented on Tracker/Screens/TrackersViewController.swift:
```diff
@@ -121,52 +116,20 @@ final class TrackersViewController: UIViewController {
-
+    private func getDayWithoutTime(date: Date) -> Date {
+        let dateWithoutTime = Calendar.current.dateComponents([.year, .month, .day], from: date)
+        return Calendar.current.date(from: dateWithoutTime)!
```
давай здесь тоже избавимся от force unwrap

LizaIlinykh commented on Tracker/Screens/TrackersViewController.swift:
```diff
@@ -247,15 +208,14 @@ extension TrackersViewController: UICollectionViewDataSource {
-        let headerView = collectionView.dequeueReusableSupplementaryView(ofKind: UICollectionView.elementKindSectionHeader, withReuseIdentifier: "header", for: indexPath) as! SupplementaryView
-        let category = visibleCategories[indexPath.section]
-        headerView.titleLabel.text = category.name
+        let headerView = collectionView.dequeueReusableSupplementaryView(ofKind: UICollectionView.elementKindSectionHeader, withReuseIdentifier: "header", for: indexPath) as! HeaderMainScreenView
```
здесь тоже не стоит использовать принудительный каст, нужно заменить на as?

MickeyRU committed:
Фиксы после ревью:
- убраны принудительные распаковки опционалов
- поправлены названия методов
- поправлены названия переменных
- поправлены модификаторы доступов
